### PR TITLE
Portal: Fix onOutClick after child re-render.

### PR DESCRIPTION
**Pull request by sompylasar**
* The previous implementation triggered `onOutClick` after a click in
the `Portal` content if it gets re-rendered during that click. It
assumed that if the clicked element is not found in the root element
via `root.contains(e.target)`, it's outside. But if after re-render the
clicked element gets removed from the DOM, so it cannot be found in the
root element. Instead we capture and flag the click event before it
bubbles up to the `document` to be handled by `handleOutClick`.

* It's recommended to use `ref` callbacks instead of `findDOMNode`.
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-find-dom-node.md

* Extra `handleOutClick` on the prototype is not needed, it is only
used if `canUseDOM` and in that case gets overwritten in the `constructor`.

### CODE REVIEW

**Comment by sunify** [comment]
Wow, thanks! I will check this in couple days and merge if all ok :-)

**Comment by sompylasar** [comment]
Please post an update when this goes to npm. Thanks.

File: src/Portal.js [hunk 1]
```diff
@@ -15,40 +15,64 @@ export default class Portal extends React.Component {
...
   }
 
-  componentWillUpdate({ onOutClick, ...props }) { // eslint-disable-line
-    this.element = ReactDOM.unstable_renderSubtreeIntoContainer(
+  componentWillUpdate({ onOutClick, ...props }) {  // eslint-disable-line no-unused-vars
+    // It's recommended to use `ref` callbacks instead of `findDOMNode`. https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-find-dom-node.md
+    ReactDOM.unstable_renderSubtreeIntoContainer(
       this,
-      <div {...props} />,
+      <div {...props} ref={this.handleRootRef} />,
```
Comment (sunify) [inline src/Portal.js]: We can use onClick prop instead dances with refs, it also works

Comment (sompylasar) [inline src/Portal.js]: React events propagate differently, always attached to document. I'd like to keep both handlers in the same boat (raw DOM events). And this solution has been tested in a real product where the bug occurred.

Comment (sunify) [inline src/Portal.js]: Ok, merged

**Comment by sunify** [comment]
New version published to npm

**Comment by sompylasar** [comment]
Thanks @sunify!

**Pull request merged**